### PR TITLE
Fixed issues #3, #4, #5

### PR DESCRIPTION
**Pull request by ZieIony**
- changed clipping method,
- changed shadow casting method,
- removed unnecessary content layout,
- added support for padding,
- refactored a bit,
- updated the main sample

### CODE REVIEW

File: app/src/main/res/layout/activity_main.xml [hunk 1]
```diff
@@ -11,9 +11,10 @@
     <com.github.florent37.diagonallayout.DiagonalLayout
         android:id="@+id/diagonalLayout"
         android:layout_width="match_parent"
-        android:layout_height="250dp"
+        android:layout_height="266dp"
         diagonal:diagonal_angle="10"
         android:elevation="10dp"
+        android:paddingBottom="16dp"
```
Comment (ZieIony) [inline app/src/main/res/layout/activity_main.xml:17]: make some space for the layout to draw its shadow inside

File: diagonallayout/src/main/java/com/github/florent37/diagonallayout/DiagonalLayout.java [hunk 2]
```diff
@@ -1,19 +1,40 @@
...
 import android.view.ViewGroup;
+import android.view.ViewOutlineProvider;
 import android.widget.FrameLayout;
-import java.util.ArrayList;
-import java.util.List;
 
 public class DiagonalLayout extends FrameLayout {
 
-    DiagonalLayoutContent content;
+    private static final float EPSILON = 0.5f;
```
Comment (ZieIony) [inline diagonallayout/src/main/java/com/github/florent37/diagonallayout/DiagonalLayout.java:22]: without an epsilon there are visible antialiasing artifacts

File: diagonallayout/src/main/java/com/github/florent37/diagonallayout/DiagonalLayout.java [hunk 3]
```diff
@@ -27,28 +48,155 @@ public DiagonalLayout(Context context, AttributeSet attrs) {
...
-    protected void onFinishInflate() {
-        super.onFinishInflate();
-        content = new DiagonalLayoutContent(getContext());
-        content.setSettings(settings);
-
-        final List<View> subViews = getAndRemoveSubViews();
-        addView(content, ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT);
-        for(View view : subViews) {
-            content.addView(view);
+    public ViewOutlineProvider getOutlineProvider() {
```
Comment (ZieIony) [inline diagonallayout/src/main/java/com/github/florent37/diagonallayout/DiagonalLayout.java:147]: the right way to specify a shadow shape

File: diagonallayout/src/main/java/com/github/florent37/diagonallayout/DiagonalLayout.java [hunk 4]
```diff
@@ -27,28 +48,155 @@ public DiagonalLayout(Context context, AttributeSet attrs) {
...
+    @Override
+    protected void dispatchDraw(Canvas canvas) {
+        int saveCount = canvas.saveLayer(0, 0, getWidth(), getHeight(), null, Canvas.ALL_SAVE_FLAG);
+
+        super.dispatchDraw(canvas);
+
+        paint.setXfermode(pdMode);
+        canvas.drawPath(clipPath, paint);
+
+        canvas.restoreToCount(saveCount);
```
Comment (ZieIony) [inline diagonallayout/src/main/java/com/github/florent37/diagonallayout/DiagonalLayout.java:199]: a trick to clip content using per-pixel operations and drawing layers

**Comment by florent37** [comment]
wow i'm impressed, you did a very great job.
I didn't know the OutlineProvider, I was looking for something like this but didn't find it

**Comment by florent37** [comment]
I merged it and pushed the 1.0.3

**Pull request merged**